### PR TITLE
Add 'Back' Button for Touch TFT 320x240

### PR DESCRIPTION
**Pull request by robbycandra**
Add "Back" / Esc Button for Touch Screen.

The button is usefull to remove "MENU_BACK" item.

### CODE REVIEW

File: Marlin/src/lcd/dogm/u8g_dev_tft_320x240_upscale_from_128x64.cpp [hunk 1]
```diff
@@ -214,76 +224,97 @@ static const uint8_t ili9341_init_sequence[] = { // 0x9341 - ILI9341
...
+     B10000000,B00011111,B11111100,B00000001,
+     B10000000,B00001100,B00000000,B00000001,
+     B10000000,B00000100,B00000000,B00000001,
+     B10000000,B00000000,B00000000,B00000001,
+     B10000000,B00000000,B00000000,B00000001,
+     B10000000,B00000000,B00000000,B00000001,
+     B10000000,B00000000,B00000000,B00000001,
+     B10000000,B00000000,B00000000,B00000001,
+     B01111111,B11111111,B11111111,B11111110,
+  };
```
Comment (tpruvot) [inline Marlin/src/lcd/dogm/u8g_dev_tft_320x240_upscale_from_128x64.cpp]: miss a line look like 
image

File: Marlin/src/lcd/dogm/u8g_dev_tft_320x240_upscale_from_128x64.cpp [hunk 2]
```diff
@@ -214,76 +224,97 @@ static const uint8_t ili9341_init_sequence[] = { // 0x9341 - ILI9341
...
+     B10001000,B00000000,B00100010,B00000001,
+     B10001000,B00000100,B00100010,B00010001,
+     B10001111,B11000011,B11000001,B11100001,
+     B10000000,B00000000,B00000000,B00000001,
+     B10000000,B00000000,B00000000,B00000001,
+     B10000000,B00000000,B00000000,B00000001,
+     B10000000,B00000000,B00000000,B00000001,
+     B10000000,B00000000,B00000000,B00000001,
+     B01111111,B11111111,B11111111,B11111110,
   };
```
Comment (tpruvot) [inline Marlin/src/lcd/dogm/u8g_dev_tft_320x240_upscale_from_128x64.cpp:251]: what about a cross (X) for ESC?

File: Marlin/src/lcd/dogm/u8g_dev_tft_320x240_upscale_from_128x64.cpp (hunk 1 above)
Comment (tpruvot) [reply Marlin/src/lcd/dogm/u8g_dev_tft_320x240_upscale_from_128x64.cpp]: i would keep "color btns left" for the arrows, the other one for ok/cancel

**Comment by tpruvot** [comment]
cant push to this PR, so i made a branch : https://github.com/tpruvot/Marlin/commits/back_button

image

File: Marlin/src/lcd/dogm/u8g_dev_tft_320x240_upscale_from_128x64.cpp [hunk 3]
```diff
@@ -332,7 +363,7 @@ uint8_t u8g_dev_tft_320x240_upscale_from_128x64_fn(u8g_t *u8g, u8g_dev_t *dev, u
     uint16_t* buffer = &bufferA[0];
     bool allow_async = true;
   #else
-    uint16_t buffer[256]; // 16-bit RGB 565 pixel line buffer
+    uint16_t buffer[WIDTH]; // 16-bit RGB 565 pixel line buffer
```
Comment (tpruvot) [inline Marlin/src/lcd/dogm/u8g_dev_tft_320x240_upscale_from_128x64.cpp]: need at least 150 for the bottom line

Comment (tpruvot) [reply Marlin/src/lcd/dogm/u8g_dev_tft_320x240_upscale_from_128x64.cpp]: need 256 in fact = 512 bytes (4 times 128 bytes u8g_WriteSequence)

**Comment by thinkyhead** [comment]
> cant push to this PR, so i made a branch

```sh
git remote add -f robbycandra git@github.com:robbycandra/Marlin.git
git checkout robbycandra/Add_Touch_Back_Button -b my_touch_changes
git push --set-upstream origin my_touch_changes
```
You can then send a PR to `robbycandra/Add_Touch_Back_Button` from your `my_touch_changes` branch. You'll see the link on your Github page after you push your changes.

https://github.com/robbycandra/Marlin/compare/Add_Touch_Back_Button...tpruvot:my_touch_changes?expand=1

**Comment by thinkyhead** [comment]
I was able to make a PR from @tpruvot's branch, so you can bring in the changes by merging robbycandra/Marlin#1.

**Comment by robbycandra** [comment]
Oops... what should i do now ?

File: Marlin/src/lcd/dogm/u8g_dev_tft_320x240_upscale_from_128x64.cpp (hunk 2 above)
Comment (robbycandra) [reply Marlin/src/lcd/dogm/u8g_dev_tft_320x240_upscale_from_128x64.cpp:251]: What if "BACK" instead of "X" ?

**Comment by tpruvot** [comment]
just merge the PR on your repo and it should appear here...

**Comment by tpruvot** [comment]
may need a bit more work then, like these color defines to rename, left/right is no more correct. ARROWS for left, OPS for right maybe

File: Marlin/src/lcd/dogm/u8g_dev_tft_320x240_upscale_from_128x64.cpp (hunk 2 above)
Comment (tpruvot) [reply Marlin/src/lcd/dogm/u8g_dev_tft_320x240_upscale_from_128x64.cpp:251]: ESC and BACK are english, we cant translate that

Comment (robbycandra) [reply Marlin/src/lcd/dogm/u8g_dev_tft_320x240_upscale_from_128x64.cpp:251]: OK.. we use "X"

**Comment by robbycandra** [comment]
I make the button size all equal. 32 x 20 pixel

**Comment by tpruvot** [comment]
there is a reason for the reduced size... on our printer, we cant reach easily the right border due to the metal shield... else your x top/bottom right has some zeros...

image

cf https://drive.google.com/file/d/11_S9CSyaMD6YE1qDI4D4xvARHiqndTvk/view?usp=sharing

**Comment by tpruvot** [comment]
to keep some margin on left/right borders (same as bottom), maybe the 2 middle buttons could be "merged" in one, without the space between

**Comment by tpruvot** [comment]
https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.0.x/config/examples/Alfawise/U20/Configuration.h#L101-L109

**Comment by tpruvot** [comment]
else better now, just the X height seems too big vs other icons...

**Comment by robbycandra** [comment]
P_20190827_114714_vHDR_Auto
P_20190827_114728_vHDR_Auto
P_20190827_114720_vHDR_Auto

**Comment by tpruvot** [comment]
ok but... these buttons are for dogm upscale mode, i dont see dogm ui there.

I mean these buttons are defined for the code where it is... and for your UI, should not be in this file..

**Comment by tpruvot** [comment]
we will see, soon or later, a problem with the hardcoded positions in the xpt touchscreen code... maybe we should do vars for that

**Comment by robbycandra** [comment]
I have made the "X" smaller. Yes, smaller "X" is better. 

Yes... Thats full scale screen. But some touch related function still not working.

**Comment by tpruvot** [comment]
to keep the colors tunable, need to apply that https://github.com/tpruvot/Marlin/commit/1d2ec10a012decf237348b3174ef58e989aa6a0c

**Comment by robbycandra** [comment]
BTW,.. how you put "SD" text at top of your screen ?

**Comment by tpruvot** [comment]
its something i didnt "PR" yet, its my master branch

https://github.com/tpruvot/Marlin/commit/712497b6c67079b6439208f692d22b684a0cc851#diff-ec2f6f0048410f90bf566ce7f98b6ef1R100

**Comment by tpruvot** [comment]
yea euh, the sample is ok, but main config change should not be like that i think... but maybe can be left so thinkyhead can move it at the right place... in TFT or touch section

**Comment by robbycandra** [comment]
I use u8g_dev_pb8v2_base_fn, This give me 4 color for my screen that i can control from ultralcd.

**Comment by tpruvot** [comment]
else, another register related to touch buttons... Marlin has some config to enable the encoder on status screen.. Rmm dont remember for what.. but i tried it and was not working on touch buttons.. something to check...

**Comment by robbycandra** [comment]
> yea euh, the sample is ok, but main config change should not be like that i think... but maybe can be left so thinkyhead can move it at the right place... in TFT or touch section

My touch screen now is working at menu text position. but my code is really a mess. 
I make many modification on ultralcd and menu. I dont think it can be easily merged to marlin.

**Comment by tpruvot** [comment]
yep, same for our top icons, its kinda too specific and messy... not optimal. SD is ok, but USB is never off once "connected" :p i just detect the first octoprint command... to turn it "on"

and FS (filament sensor) is more a RED flag when the error happen

**Comment by robbycandra** [comment]
@tpruvot , this is my touch code...

touch class.
```
  if (!isTouched()) return 0;
  const uint16_t x = uint16_t(((uint32_t(getInTouch(XPT2046_X))) * tsoffsets[0]) >> 16) + tsoffsets[1],
                 y = uint16_t(((uint32_t(getInTouch(XPT2046_Y))) * tsoffsets[2]) >> 16) + tsoffsets[3];
  if (!isTouched()) return 0; // Fingers must still be on the TS for a valid read.

  const uint8_t row_touched = y / 16;
  const uint8_t col_touched = x / 40;

  if (row_touched < 12) {
    return (128 + (row_touched << 3) + col_touched);
  } 
  else {
    return  WITHIN(x,  14,  77) ? EN_D
          : WITHIN(x,  90, 153) ? EN_A
          : WITHIN(x, 166, 229) ? EN_B
          : WITHIN(x, 242, 305) ? EN_C
          : 0;
  }
```

read touched area.
```
    bool MarlinUI::menu_is_touched(int8_t tested_item_number) {
      uint8_t touched_item_number;
      bool menu_area_touched = false;
      if (lcd_menu_touched_coord & B10000000) {
        const uint8_t row = (lcd_menu_touched_coord & B01111000) >> 3;
        const uint8_t col = (lcd_menu_touched_coord & B00000111); 
        if (ui.menu_mode == MENU_2X4) {
          touched_item_number = (int)(row / 3) * 2 + (col >> 2);
          menu_area_touched = true;
        }
        else if (ui.menu_mode == MENU_1X6) {
          touched_item_number = row >> 1;
          menu_area_touched = true;
        }
        else if (row > 8) {  // ui.menu_mode = MENU_SELECT & 4th row.
          touched_item_number = col >> 2; 
          menu_area_touched = true;
        }
        if (menu_area_touched) {
          if (touched_item_number == tested_item_number) {
            lcd_menu_touched_coord = 0;
            #if HAS_BUZZER
              buzz(LCD_FEEDBACK_FREQUENCY_DURATION_MS, LCD_FEEDBACK_FREQUENCY_HZ);
            #endif
            return true;
          }
        }
      } 
      return false;
    }

```

**Comment by tpruvot** [comment]
TFT_TOPICONS_COLOR define should be removed from configs for now :p else double check the default colors, was just "random" values as sample colors

**Comment by robbycandra** [comment]
OK, Removed

**Comment by tpruvot** [comment]
current (uncommented) config values : 

image

**Comment by robbycandra** [comment]
> current config values :

I think its OK.

File: config/examples/Alfawise/U20/Configuration.h [hunk 4]
```diff
@@ -101,13 +101,6 @@
 // 4 - If you want to tune the UI colors, define custom ones here. RGB 16 bits 5-6-5 format
 // see https://ee-programming-notepad.blogspot.com/2016/10/16-bit-color-generator-picker.html
```
Comment (tpruvot) [inline config/examples/Alfawise/U20/Configuration.h]: these 2 lines need to be removed...

File: config/examples/Alfawise/U20/Configuration.h [hunk 5]
```diff
@@ -2142,9 +2134,18 @@
 //=============================================================================
 
 //
-// FSMC display (MKS Robin, Alfawise U20, JGAurora A5S, A1, etc.)
+// FSMC display (MKS Robin, Alfawise U20, JGAurora A5S, REXYZ A1, etc.)
 //
-#define FSMC_GRAPHICAL_TFT
+//#define FSMC_GRAPHICAL_TFT
```
Comment (tpruvot) [inline config/examples/Alfawise/U20/Configuration.h]: you disabled our TFT

File: config/examples/Alfawise/U20/Configuration.h (hunk 4 above)
Comment (robbycandra) [reply config/examples/Alfawise/U20/Configuration.h]: Done.

File: config/examples/Alfawise/U20/Configuration.h (hunk 5 above)
Comment (robbycandra) [reply config/examples/Alfawise/U20/Configuration.h]: Done

**Comment by tpruvot** [comment]
good to go for me else... maybe the colors should stay commented by default, unsure

**Comment by gustavx404** [comment]
you could help me make the display of my etet et4 look similar to this mine doesn't appear anything is blank i have a skr 1.3 and marlin 2 bug fix i don't know how to add this panel, sorry for the translation i used google translator

**Pull request merged**